typing/typemod.ml

```diff
@@ -1647,26 +1761,52 @@ and type_structure ?(toplevel = false) funct_body anchor env sstr scope =
   let rec type_struct env sstr =
     Ctype.init_def(Ident.current_time());
     match sstr with
-    | [] -> ([], [], env)
-    | pstr :: srem ->
+    | [] -> ([], [], [], env)
+    | pstr :: srem -> begin
         let previous_saved_types = Cmt_format.get_saved_types () in
         let desc, sg, new_env = type_str_item env srem pstr in
         let str = { str_desc = desc; str_loc = pstr.pstr_loc; str_env = env } in
         Cmt_format.set_saved_types (Cmt_format.Partial_structure_item str
                                     :: previous_saved_types);
-        let (str_rem, sig_rem, final_env) = type_struct new_env srem in
-        (str :: str_rem, sg @ sig_rem, final_env)
+        let (str_rem, sig_rem, fsig_rem, final_env) = type_struct new_env srem in
+        match extract_open_struct desc with
+        | Some (tm, md, id, md_env) -> begin
+            let loc = pstr.pstr_loc in
+            let tm_str = {str_desc = tm; str_loc = loc;
+                          str_env = env} in
+            let open_str = {str with str_env = md_env} in
+            let md_sig =
+              Sig_module (id, {md_type=md.Types.md_type; md_loc=loc;
+                               md_attributes = []}, Trec_not) in
+            if not (in_nested_struct ()) then begin
```

Comment:
Can you explain the logic behind in_nested_struct?

Reply:
The following program should not be rejected.

```OCaml
module M = struct open struct type t = T let x = T end end;;
```

Reply:
The naming of `in_nested_struct` is misleading. I will fix it.